docs/pages/guides/onchain-indices.mdx

```diff
@@ -0,0 +1,412 @@
+
+For this reason, we create onchain “indices” that maintain the result of these transformations.
+
+## Indices in MUD
+
+In these examples, we will make indices for a simple game where each account has an (x, y) position.
+
+```ts filename="mud.config.ts" copy showLineNumbers
+import { mudConfig } from "@latticexyz/world/register";
+
+export default mudConfig({
+  tables: {
+    Position: {
+      keySchema: {
+        player: "address",
+      },
+      valueSchema: {
+        x: "int32",
+        y: "int32",
+      },
+    },
+  },
+});
+```
+
+In this game, players update their position by calling `MoveSystem.move` with their desired coordinates.
+
+```solidity filename="MoveSystem.sol" copy showLineNumbers
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.24;
+
+import { System } from "@latticexyz/world/src/System.sol";
+import { Position } from "../codegen/index.sol";
+
+contract MoveSystem is System {
+  function move(int32 x, int32 y) public {
+    address player = _msgSender();
+
+    Position.set(player, x, y);
+  }
+}
+```
+
+To make the game more interesting, we would like to prevent players from moving to positions that are already occupied. However, it's not clear how to tell which players are at a given position onchain.
+
+   <CollapseCode>
+
+```solidity filename="MoveSystem.sol" copy showLineNumbers {11-14}
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.24;
+
+import { System } from "@latticexyz/world/src/System.sol";
+import { Position } from "../codegen/index.sol";
+
+contract MoveSystem is System {
+  function move(int32 x, int32 y) public {
+    address player = _msgSender();
+
+    bool positionIsEmpty = ?
+    require(positionIsEmpty, "another player is at this position");
+
+    Position.set(player, x, y);
+  }
+}
+```
+
+   </CollapseCode>
+
+In other words, there is no reverse mapping - we can read the position of a given player, but not which players are at each position. It's time to add an index!
+
+There are two main ways to implement this - you can **create and manage your own indices**, or **use a module**.
+
+### Creating your own indices
+
+#### `PlayersAtPosition`
+
+One way to know whether a position is occupied is to have a list of which players are there, and check if that list is empty.
+
+First, we declare an `PlayersAtPosition` table in our MUD config. The schema is essentially the reverse of `Position`; the key is an (x, y) coordinate and the value is an array of addresses currently at that position:
+
+   <CollapseCode>
+```ts filename="mud.config.ts" copy showLineNumbers {14-17}
+import { mudConfig } from "@latticexyz/world/register";
+
+export default mudConfig({
+  tables: {
+    Position: {
+      keySchema: {
+        player: "address",
+      },
+      valueSchema: {
+        x: "int32",
+        y: "int32",
+      },
+    },
+    PlayersAtPosition: {
+      keySchema: { x: "int32", y: "int32" },
+      valueSchema: { players: "address[]" },
```

Comment:
```suggestion
      valueSchema: { players: "address" },
```

We are changing the game so at most one player can be at a location. If we use `address` here, it makes the next code sample *a lot* simpler.

Reply:
Interesting point. I totally agree that the `filter` is needlessly confusing. 

The code will need to be modified a bit to check if this is zero but I will try it out!

Reply:
The main reason I implemented `address[]` was to compare the app-specific index with `keysWithValue` on more equal footing - the difference is gas would come from the hooks etc. But I guess that says something that we can optimize this index based on our app.